### ISSUE TITLE
[css-inline]  `text-box-edge: ideographic` and `ideographic-ink` should use ideographic fonts in the cascade list

### ISSUE THREAD

**Issue by kojiishi** [w3c/csswg-drafts#10928]
Many CJK sites use Latin fonts as the first available fonts, to use the Latin fonts for Latin characters. For example:
```css
font-family: Arial, CJK-font, sans-serif;
```

Authors want to align text very accurately when applying `ideographic` or `ideographic-ink` for the [`text-box-edge` property](https://drafts.csswg.org/css-inline-3/#propdef-text-box-edge). In this case, using the metrics from `Arial` doesn't help, but we don't have a mechanism to use metrics from the `CJK-font`.

Can we define, at least for `text-box-edge: ideographic*`, it should use the "first available ideographic font", instead of "first available font"? How to define the "first available ideographic font" isn't technically a simple work though.

(I mentioned this in #10850 but splitting as it's has enough complexity to discuss separately)

@fantasai @nt1m @vitorroriz @Clqsin45 @bfgeek @jfkthame @kidayasuo @xfq

**Comment by jfkthame** [comment]
I'm a bit uneasy about this, as I'm worried that determining the "first available ideographic font" may be quite tricky in the general case -- issues like on-demand webfont loading and incremental font transfer may complicate things here.

For authors who want to such accurate alignment, wouldn't a better approach be to specify their desired (CJK) font as the first in the font-family list, so that its metrics are clearly the ones to use? If they want to avoid the use of Latin glyphs from that font, Latin codepoints could be excluded using `unicode-range` so that a subsequent Latin font will be used instead. Listing Arial first when the content is primarily CJK (and layout is supposed to be based on the CJK metrics) seems like the wrong technique.

**Comment by yisibl** [comment]
>  If they want to avoid the use of Latin glyphs from that font, Latin codepoints could be excluded using `unicode-range` so that a subsequent Latin font will be used instead.

@jfkthame Can you write an example of the exact way this works?

**Comment by jfkthame** [comment]
E.g. if I wanted to use "Heiti SC" for Chinese characters, but "Times New Roman" for Latin letters within the Chinese content, I could try something like
```
@font-face {
  font-family: heiti-cjk-only;
  src: local("Heiti SC");
  unicode-range: U+3000-9FFF, U+F900-FAFF; /* extend as needed */
}
body {
  font-family: heiti-cjk-only, "Times New Roman", serif;
}
```

I've put an example (should work on both Windows and macOS) at https://codepen.io/jfkthame/pen/zYgYBrE?editors=1100, showing the contrast between using the Latin characters that are part of the CJK font vs having them fall back to a chosen Latin font instead.

**Comment by yisibl** [comment]
This looks like a bit of a hack, since `unicode-range` does not currently have a syntax that excludes certain Unicode code points, which should theoretically be used here: `unicode-range: exclude(latin)`. 
Otherwise, it's hard for the author to define what is Chinese, so it's best to have an exclusion syntax where the rest of the characters are looked up from the defined font.

Also, does this affect performance?

**Comment by Crissov** [comment]
#4573 would help, though exclusions may be a nice addition.

**Comment by jfkthame** [comment]
> This looks like a bit of a hack, since `unicode-range` does not currently have a syntax that excludes certain Unicode code points, which should theoretically be used here: `unicode-range: exclude(latin)`.
> 
> Otherwise, it's hard for the author to define what is Chinese, so it's best to have an exclusion syntax where the rest of the characters are looked up from the defined font.

Agreed; more flexible and expressive unicode-range options would be nice for many reasons.

> Also, does this affect performance?

In principle it probably improves performance compared to doing the `font-family: Arial, CJK-font, sans-serif;` pattern for text that is primarily CJK, because it avoids making the browser first check Arial for each CJK character (only to find that it is not supported there). In practice the difference may well be pretty insignificant.

**Comment by yisibl** [comment]
Put another way, do we make it possible for authors to specify a font that is used as an ideographic indicator?

e.g. `preferred-font-family: CJK-font`?

**Comment by kojiishi** [comment]
> I'm a bit uneasy about this, as I'm worried that determining the "first available ideographic font" may be quite tricky in the general case -- issues like on-demand webfont loading and incremental font transfer may complicate things here.

I agree determining if a font is an ideographic font or not is very tricky, but for this specific case, it's not that tricky depends on #10850. The current spec requires the `BASE` table, so it should be straightforward to find the first font that has the specified tags in the `BASE` table.

If #10850 is going to synthesize these metrics from some glyphs, it's also feasible--the first font that has the glyphs.

If #10850 is going to some other methods, I don't know how else we can synthesize, but we can discuss.

**Comment by fantasai** [comment]
We could consider, in the same way that we take the "first available font" from the first font with a space character, taking the "first available ideographic font" from the first font with “水” (CJK water ideograph, U+6C34); and use that both for text-box-edge and the `ic` unit.

**Comment by jfkthame** [comment]
Depending on the presence of a specific ideograph may have surprising results in the presence of font subsetting (perhaps especially if it's dynamic subsetting based on varying content) or incremental fonts.

(I'm uneasy about its use to define `ic` for the same reason; I feel that may be a mistake that we'll come to regret.)

> in the same way that we take the "first available font" from the first font with a space character

[That's not quite what we do](https://drafts.csswg.org/css-fonts/#first-available-font): we take the first font for which `unicode-range` does not exclude `<space>`. This says nothing about whether it actually has a space glyph.

That avoids forcing the browser to download resources just to check their character coverage, but it's not a useful approach for the ideographic case.

**Comment by kojiishi** [comment]
I generally agree with @jfkthame, CJK fonts often use dynamic subsetting. I don't have a good idea for the `ic` unit, but for `text-box-trim`, we know the actual text. If we want to use the glyph bounding boxes, I think we can take them from the actual text and actual used fonts, similar to what `initial-letter` does?

It can't distinguish between `ideographic` and `ideographic-ink` though. And it may be useful for other scripts?

**Comment by kojiishi** [comment]
@kidayasuo [commented](https://github.com/w3c/csswg-drafts/issues/10850#issuecomment-2623768438):
> It goes without saying that these values must be calculated based on the first CJK font in the font list because the first font in the list is typically chosen based on the author’s intent for Latin glyphs rather than CJK text. This remains true even in cases where CJK is the primary script in an internationalized environment.

**Comment by kidayasuo** [comment]
The first font in the list does not necessarily reflect the author’s intention for the main font. The list is often used to override parts of the main font by placing other fonts in front of it. A common example is placing a Latin font before the main CJK font, a practice that is nearly universal in multilingual typography. That is why you need to look for the first CJK font in the list to measure the intended dimensions of the CJK text.
